bolt-modules/boltlib/lib/puppet/functions/run_task_with.rb

```diff
@@ -0,0 +1,179 @@
+                     .transform_keys { |k| k.sub(/^_/, '').to_sym }
+
+    options[:description] = description if description
+
+    # Get all the targets
+    targets = Array(inventory.get_targets(targets))
+
+    # Map the targets to their specific parameters
+    target_mapping = targets.each_with_object({}) do |target, mapping|
+      params = yield(target)
+
+      unless params.is_a?(Hash)
+        raise with_stack(
+          :TYPE_MISMATCH,
+          "Block must return a Hash of parameters, received #{params.class} for target #{target.safe_name}"
+        )
+      end
+
+      mapping[target] = params
+    end
+
+    # If all targets use the 'pcp' transport, use a fake task instead of loading the local definition
+    # Otherwise, load the local task definition
+    if targets.any? && targets.all? { |t| t.transport == 'pcp' }
+      task = Bolt::Task.new(task_name, {}, [{ 'name' => '', 'path' => '' }])
+    else
+      task_signature = Puppet::Pal::ScriptCompiler.new(closure_scope.compiler).task_signature(task_name)
+
+      if task_signature.nil?
+        raise Bolt::Error.unknown_task(task_name)
+      end
+
+      task = Bolt::Task.from_task_signature(task_signature)
+
+      # Retrieve the default parameters for the task and merge them with the
+      # target-specific parameters
+      default_params = task.parameter_defaults
+      target_mapping.transform_values! { |params| default_params.merge(params) }
+
+      target_mapping.each_value do |params|
+        task_signature.runnable_with?(params) do |mismatch_message|
+          raise with_stack(:TYPE_MISMATCH, mismatch_message)
+        end || (raise with_stack(:TYPE_MISMATCH, 'Task parameters do not match'))
+      end
+    end
+
+    target_mapping.each_value do |params|
+      # Generate a helpful error message about the type mismatch between the type Data
+      # and the actual type of params
+      unless Puppet::Pops::Types::TypeFactory.data.instance?(params)
+        params_t = Puppet::Pops::Types::TypeCalculator.infer_set(params)
+        desc = Puppet::Pops::Types::TypeMismatchDescriber.singleton.describe_mismatch(
+          'Task parameters are not of type Data. run_task()',
+          Puppet::Pops::Types::TypeFactory.data, params_t
+        )
+        raise with_stack(:TYPE_NOT_DATA, desc)
+      end
+
+      # Wrap parameters marked with '"sensitive": true' in the task metadata with a
+      # Sensitive wrapper type. This way it's not shown in logs.
+      if (param_spec = task.parameters)
+        params.each do |k, v|
+          if param_spec[k] && param_spec[k]['sensitive']
+            params[k] = Puppet::Pops::Types::PSensitiveType::Sensitive.new(v)
+          end
+        end
+      end
+
+      # executor.noop is set when run task is called from the CLI
+      # options[:noop] is set when it's called from a plan
+      if executor.noop || options[:noop]
+        if task.supports_noop
+          params['_noop'] = true
```

Comment:
What `params` is this modifying? Shouldn't this get folded into the params for each target?

Reply:
It does get folded into the params for each target. This is in a loop that iterates over each parameter hash in `target_mapping`.

Reply:
Updated to make it clearer that it updates the params mapped to each target.